### DISCUSSION TITLE
Optimistic side effects

### DISCUSSION THREAD

**Discussion by erquhart** [urql-graphql/urql#2549] [Q&A] [answered]
I have mutations that change numeric values, and on the backend, event triggers that then increment a separate numeric value based on that mutation. Both values are queried into the app in different places, and I'd like to optimistically update both, as 99.9% of the time, the incremented value can be correctly calculated.

Optimistically updating for the mutation is straightforward. For the other value, I've tried updating the cache, but cache updates under `optimistic` don't seem to be treated any differently than those in `updates`. Is there a way to properly implement optimistic updates for these kinds of side effects? The cache update API is the right tool, but the updates aren't treated as optimistic.

**Comment by erquhart** [comment]
To put it more simply - is it possible to optimistically update more than just the selection set of a mutation?

**Comment by erquhart** [comment, marked as answer]
So all updates that occur in the `optimistic` config _are_ being applied optimistically, I missed that the actual result was coming back and eliminating those updates, as it should, and the backend operations are running subsequently. This felt unexpected because the result doesn't actually impact the optimistic result from a data structure perspective, but again, urql _has_ to get rid of optimistic updates when the real result comes back, I'm just doing weird things with it. Going to try writing an updater for that mutation result so the cache maintains the optimistic result.

**Reply by wernst** [reply]
@erquhart did you end up finding a pattern for keeping around the side effects of optimistic updates when the response comes back from the network?

I'm facing a similar situation as you described

**Reply by kitten** [reply]
Hiya 👋
The easiest way to ensure this works as intended is to ensure that the optimistic update function is as sideeffect free as possible, while an updater is then used to do the sideeffect.

Applying normalised updates happens automatically, so I'm assuming what you're updating is not contained within the mutation's result.
Updater functions run regardless of whether they're operating on "real data" that's been from a mutation agains the API, or "optimistic"/fake data that's been returned by your `optimistic` function.

This means that you're basically aiming to treat the update the same. Any update that you want to preserve past the optimistic update should be reapplied by the updater. You can think of this like you were doing the same thing twice to the same state.

In the offchance you then need to also differentiate between the two, you can access `info.optimistic` in the updater, which is a flag whose only purpose is to signify whether you're doing a real or a fake update

**Reply by erquhart** [reply]
@kitten thanks for the reply! I did discover this approach, but there was still always a gap when the mutation result returned.

Example:
1. Apply optimistic update (with no side effects)
2. An updater receives the optimistic update and applies side effects
3. The mutation result returns, and both the optimistic update and side effects from the updater are eliminated (note that the side effect data did not conflict or overlap with the return data)
4. For a second or so, the mutation result with no side effects are shown in the UI
5. The updater re-applies side effects, and the UI is again accurate

This happens with complete consistency, so it seems to work as designed. If you see anything I'm missing there, please let me know!

**Reply by erquhart** [reply]
Also, I've used debugging with the client to observe where the gap occurs - it's not in rendering, it's between fetchSuccess and the updater itself running. Once the updater reapplies the update, the UI change appears synchronous, no delays.
